sbol_utilities/package.py

```diff
@@ -64,107 +65,186 @@ def regularize_package_directory(dir: str):
+            will become the namespace of the package object. Example:
+            sub-packages with namespaces 
+            "https://example.org/MyPackage/promoters" and 
+            "https://example.org/MyPackage/regulatory/repressors" would return 
+            a prefix "https://example.org/MyPackage/"
+    """
+    # Get a list of all of the namepsaces
+    all_namespaces = [package.namespace for package in subpackage_list]
+
+    # Take first word from array as reference
+    ref_string = all_namespaces[0]
+    ref_string_length = len(ref_string)
+
+    # Make a holder for the prefix
+    prefix = ""
+
+    # Start with the first letter of the reference sub-string and add one letter
+    # each loop
+    for i in range(1, ref_string_length+1):
+        # Generating the substring
+        stem = ref_string[0:i]
 
-def check_namespaces(doc: sbol3.Document):
-    """ Check if the namespaces of all top level objects are the same
+        # Check all the namespaces have the substring
+        is_present = all(stem in namespace for namespace in all_namespaces)
+        
+        if not is_present:
+            break
+
+        # If current substring is present in all strings and its length is 
+        # greater than current result, save it as the stem
+        if (is_present and len(prefix) < len(stem)):
+            prefix = stem
+
+    # Check if the namespace contains a .com or .org
+    if any(s in prefix for s in ('.com', '.org', '.edu')):
+        return(prefix)
+    else:
+        raise ValueError(f'The longest prefix found was {prefix}, which does '
+                         f'include a .com. .org, or .edu')
+
+def check_namespaces(package, sub_package_list=None):
+    """ Check if the namespaces of all top level objects in a defined package
+    are the same, and if all of its subpackages have the pacakge namespace as a
+    stem and are valid packages on their own
 
     Args:
-        doc (sbol3.Document): Document containing top level objects
+        package (sep054.Package): Package to check
+        sub_package_list (list of sep054.Package): The list of package objects 
+            correlating to the URIs listed in package.subpackages
 
     Returns:
         is_package (boolean): True if all namespaces are the same
-        namespace (string): Namespace of the package
     """
-
+    # Check all members of the root package have the same namespace
     # Get a list of all namespaces
-    all_namespaces = [o.namespace for o in doc.objects]
+    # To get a namespace, remove the object name from the URI for each member
+    all_namespaces = ["/".join(URI.split('/')[0:-1]) for URI in package.members]
 
     # Check all namespaces are the same
-    is_package = all(x == all_namespaces[0] for x in all_namespaces)
-
-    # Get the first namespace to pass the string
-    # Which position you pick won't matter if it is a package
-    namespace = all_namespaces[0]
+    if all(x == all_namespaces[0] for x in all_namespaces):
```

Comment:
Note that another, slightly more elegant way to do this, is to have all_namespaces be generated as a set, rather than a list. Using a set will automatically prune duplicates. Then you can just check if the len(all_namespaces)==1.

Reply:
Added the check with a set in commit hgscott/SBOL-utilities@312854b3969d46a59f34f4ee11cd5bd2c352da0b